[FIX] hr_expense: set main attachment on `hr.expense.sheet`

pivi-odoo opened the pull request:
## Issue
Main attachments on expense sheets was never correctly set, leading to missing attachment previews in the hr.expense.sheet form view, and unnecessary RPC requests to `register_main_attachment`, because the attachment was never set.

## Steps to reproduce
- Install Expenses
- Create 2 different expenses with different attachments
- Create a report from those 2 expenses, save it
- Notice there is nothing in the preview for the attachments, but if you zoom in or zoom out the page, then you can see it. Also swiping on the list of attachments isn't persistent, once you refresh the page, the main attachment we selected lastly is lost.

## Cause
There are 2 main issues:
- `hr.expense.sheet` has `message_main_attachment_id`, but it's never set.
- When writing an attachment via `register_main_attachment`, we pass only the id of the attachment. Then in the backend, the related record is based on the model of the attachment, but the attachment is linked to `hr.expense`, not `hr.expense.sheet`. Therefor we re-write the same attachment on the record we read the attachment from. And then the frontend will continue to call `register_main_attachment` everytime it tries to find the attachments linked to the sheet, but it's never done.

## Fix
- Add a compute to initially set the value of the `message_main_attachment_id` based on the spec "we take the first line that has an attachment and we set it on the sheet"
- The `hr.expense.sheet` now maintains a copies of the attachments linked to it's expenses, for proper ACL and functionality of `register_main_attachment`.

## Affected versions
16.0 up to master = saas-17.1

## Reference
task-3572440

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/142029).

pivi-odoo commented:
Hi @xmo-odoo, thanks for your review.

>Because the attachment is not copied over, the access checks will not be correctly performed and the relation is inconsistent (the main attachment of the sheet is not in any way related to a sheet let alone this one)

I've reworked completely the patch to follow your suggestion, I've had a bit of a hard time maintaining the sync between the attachments of the expenses and those on the sheet, so I decided to use the `checksum` of the attachments to avoid unnecessary trashing. If you see a better solution, I am all for it.

Let me know what you think.

JulienVR commented on addons/hr_expense/models/hr_expense.py:
```diff
@@ -1070,6 +1108,19 @@ def create(self, vals_list):
+
+    def unlink(self):
+        self._unlink_sheet_attachments()
+        return super(HrExpenseSheet, self).unlink()
```
you can just do `super().unlink()`

JulienVR reviewed (approved):
It's good to me. Y ou'll just have to ask a runbot exception to the upgrade team to make the CI happy I think. :)

pivi-odoo commented:
Hi @JulienVR,

The Runbot's template exception has been granted, it just need the r+ :)

JulienVR commented:
I don't have r+ rights :sweat:, but @csnauwaert do (I'm only pre-reviewing PRs)

seb-odoo reviewed (commented):
Hello, some comments as I'm passing by.

Also, as this is rather tricky, we should have some tests to cover this behavior.

pivi-odoo commented on addons/hr_expense/models/hr_expense.py:
```diff
@@ -243,6 +244,10 @@ def _compute_attachment_number(self):
             expense.attachment_number = attachment.get(expense._origin.id, 0)
 
+    def _compute_attachment_ids(self):
+        for expense in self:
+            expense.attachment_ids = self.env['ir.attachment'].search([('res_model', '=', 'hr.expense'), ('res_id', '=', expense.id)])
```
It's forbidden to do a aggregate on the ids of the attachments by the ORM. (I presume it's for some security reason, or to not bypass the ACL).

I've tried initially to do a `_read_group` and array_agg(id) and was met with the error.

I know it's not ideal, but if I can't agg on ids, I don't think I can do better? Unless I am missing something?

ryv-odoo replied:
Isn't this equivalent to a One2many with a domain, or I miss something ?
```python
attachment_ids = fields.One2many('ir.attachment', 'res_id', domain=[('res_model', '=', 'hr.expense')], string="Attachments")
```
(It will be batched automatically by the ORM)

pivi-odoo commented on addons/hr_expense/models/hr_expense.py:
```diff
@@ -970,6 +977,37 @@ def _compute_attachment_number(self):
+            return (attachments_lhs.filtered(lambda att: att.checksum in diff and att.checksum not in checksums_rhs),
+                    attachments_rhs.filtered(lambda att: att.checksum in diff and att.checksum not in checksums_lhs))
+
+        for sheet in self:
+            sheet_attachments = self.env['ir.attachment'].search([('res_model', '=', 'hr.expense.sheet'), ('res_id', '=', sheet.id)])
```
ditto from my comment above

ryv-odoo commented on addons/hr_expense/models/hr_expense.py:
```diff
@@ -970,6 +977,37 @@ def _compute_attachment_number(self):
+        for sheet in self:
+            sheet_attachments = self.env['ir.attachment'].search([('res_model', '=', 'hr.expense.sheet'), ('res_id', '=', sheet.id)])
+            attachment_to_copy, attachment_to_unlink = diff_attachments(sheet.expense_line_ids.attachment_ids, sheet_attachments)
+            for attachment in attachment_to_copy:
+                sheet_attachments += attachment.copy({
+                    'res_id': sheet.id,
+                    'res_model': 'hr.expense.sheet',
+                })
+            sheet_attachments -= attachment_to_unlink
+            attachment_to_unlink.unlink()
```
Indeed, I strongly suggest to do it in C(RU)D instead. The compute can be legitimately be called by `onchange` on unsaved/new records.

pivi-odoo commented:
@JulienVR @csnauwaert I've updated the PR, to apply @seb-odoo and @ryv-odoo suggestions of moving the syncing of the attachments to the CRUD operations. I've also added a test that covers the use-cases that I could think of.

Since I guess the previous approval is considered outdated, since the patch completely changed.

Since I'm extending ir.attachment now to override the create and unlink, I guess I need to reping @odoo/rd-security (@xmo-odoo) but I think it's ok.

pivi-odoo commented on addons/hr_expense/models/hr_expense.py:
```diff
@@ -454,6 +457,14 @@ def write(self, vals):
                     self.sheet_id.write({'employee_id': vals['employee_id']})
                 elif len(employees) > 1:
                     self.sheet_id = False
+        if sheet_ids or self.sheet_id:
+            (sheet_ids | self.sheet_id)._sync_attachments_with_expenses()
```
sheet_ids may contain multiple sheets (since self can contain multiple expenses, each associated with different sheets).

I will correct the case for no-op case (the call does nothing, but I guess it's call can be avoided)

pivi-odoo commented on addons/hr_expense/models/hr_expense.py:
```diff
@@ -1059,6 +1078,30 @@ def _search_product_ids(self, operator, value):
+                    attachments_rhs.filtered(lambda att: att.checksum in diff and att.checksum not in checksums_lhs))
+
+        for sheet in self:
+            attachment_to_copy, attachment_to_unlink = diff_attachments(sheet.expense_line_ids.attachment_ids, sheet.attachment_ids)
```
True, I will contact a PO, because I am not sure anymore what was the initial idea for the attachments of expense sheets.

pivi-odoo commented on addons/hr_expense/models/hr_expense.py:
```diff
@@ -1059,6 +1078,30 @@ def _search_product_ids(self, operator, value):
+            """
+            checksums_lhs = set(attachments_lhs.mapped('checksum'))
+            checksums_rhs = set(attachments_rhs.mapped('checksum'))
+            diff = checksums_lhs ^ checksums_rhs
+            return (attachments_lhs.filtered(lambda att: att.checksum in diff and att.checksum not in checksums_rhs),
+                    attachments_rhs.filtered(lambda att: att.checksum in diff and att.checksum not in checksums_lhs))
```
True, it's equivalent, the `diff` can be omitted.

pivi-odoo commented:
Ok, so I've contacted the PO of expenses to get the full specs after describing the current issue and the consensus was:
- Full sync between attachments from expenses and those found on the report. Which means that if attachments on the expense change, those on the report should be adjusted too, and *vice-versa*, changing an attachment on the report, that originally came from one of the expenses, should also change the attachment on the expense.
- Report can have their own attachments, they shouldn't be affected by the syncing process.

So this is the third iteration of the fix (hoping last '-.-), sadly because of the need to not modify attachments that are on the report, I didn't find a clean abstraction for the syncing process, so for now it's a bit ad-hoc :S (I am all ears for any suggestions)

ryv-odoo commented on addons/hr_expense/models/hr_expense.py:
```diff
@@ -87,6 +87,7 @@ def _get_employee_id_domain(self):
         ("company_account", "Company")
     ], default='own_account', tracking=True, states={'done': [('readonly', True)], 'approved': [('readonly', True)], 'reported': [('readonly', True)]}, string="Paid By")
     attachment_number = fields.Integer('Number of Attachments', compute='_compute_attachment_number')
+    attachment_ids = fields.One2many('ir.attachment', 'res_id', domain="[('res_model', '=', 'hr.expense')]", string="Attachments")
```
I think that it is already done by the magic `_search` override of `ir.attachement` 

https://github.com/odoo/odoo/blob/ad702c20b127b8ae4d28ba6bbd61b31ad6074808/odoo/addons/base/models/ir_attachment.py#L525

JulienAlardot commented:
@pivi-odoo Take note that task-3572440 in 17.0 is reworking the main_attachment_id of expenses & its display on reports
Haven't read the code yet but
> Add a compute to initially set the value of the message_main_attachment_id based on the spec "we take the first line that has an attachment and we set it on the sheet"

Should be 
> Add a compute to initially set the value of the message_main_attachment_id based on the spec "we take ***the first expense line message_main_attachment_id*** and we set it on the sheet"

JulienAlardot commented on addons/hr_expense/models/hr_expense.py:
```diff
@@ -454,8 +458,28 @@ def write(self, vals):
                     self.sheet_id.write({'employee_id': vals['employee_id']})
                 elif len(employees) > 1:
                     self.sheet_id = False
+        if 'sheet_id' in vals:
+            for expense in self:
+                prev_sheet = expense_to_prev_sheet[expense]
+                checksums = set((expense.attachment_ids - prev_sheet.expense_line_ids.attachment_ids).mapped('checksum'))
+                prev_sheet.attachment_ids.filtered(lambda att: att.checksum in checksums).with_context(sync_attachment=False).unlink()
+                if vals['sheet_id']:
+                    if expense.sheet_id != prev_sheet:
+                        for attachment in expense.attachment_ids.with_context(sync_attachment=False):
+                            attachment.copy({
+                                'res_model': 'hr.expense.sheet',
+                                'res_id': vals['sheet_id'],
+                            })
```
A comment for this one would be nice :)
Also, nitpick but "prev" instead of "previous" isn't saving a lot of characters and less clear imo

pivi-odoo replied:
I've added a few comments to explain, it's true it's the most tricky part of the patch, let me know if something isn't clear.
Also renamed prev -> previous, since it's true that prev could be misinterpreted as "preview" for ex.

JulienAlardot commented on addons/hr_expense/models/hr_expense.py:
```diff
@@ -454,8 +458,28 @@ def write(self, vals):
+                checksums = set((expense.attachment_ids - prev_sheet.expense_line_ids.attachment_ids).mapped('checksum'))
+                prev_sheet.attachment_ids.filtered(lambda att: att.checksum in checksums).with_context(sync_attachment=False).unlink()
+                if vals['sheet_id']:
+                    if expense.sheet_id != prev_sheet:
```
A reason why this isn't higher in the tree?
It feels like it could unlink to copy the same one in weird corner-cases

pivi-odoo replied:
>A reason why this isn't higher in the tree?

Since we fall in this case not only when we write with `sheet_id` on the expense itself, but also when processing commands on the `expense_line_ids` of the sheet itself, I needed to handle all this after the `super`.
I need to know what are the attachments that will be left on the previous sheet so I know what to sync.

>It feels like it could unlink to copy the same one in weird corner-cases

Normally the no-op case is covered correctly: for the deletion we substract the attachments on the expense lines of the previous sheet, this prevents deleting attachments when we write with the same sheet_id as the current expense.
For the copy, we just check that the new sheet of the expense is != than the old one, this will cover potential duplication.

If you see a corner case I don't see, let me know, it's a tricky segment :S

JulienAlardot commented on addons/hr_expense/models/hr_expense.py:
```diff
@@ -454,8 +458,28 @@ def write(self, vals):
+    def unlink(self):
+        for sheet in self.sheet_id:
+            checksums = set((sheet.expense_line_ids.attachment_ids & self.attachment_ids).mapped('checksum'))
+            sheet.attachment_ids.filtered(
+                lambda att: att.checksum in checksums
+            ).with_context(sync_attachment=False).unlink()
```
+1

JulienAlardot commented on addons/hr_expense/models/hr_expense.py:
```diff
@@ -927,6 +927,8 @@ def _default_bank_journal_id(self):
     currency_id = fields.Many2one('res.currency', string='Currency', states={'draft': [('readonly', False)]},
                                   compute='_compute_currency_id', store=True, readonly=True)
     attachment_number = fields.Integer(compute='_compute_attachment_number', string='Number of Attachments')
+    message_main_attachment_id = fields.Many2one(compute='_compute_main_attachment', store=True)
```
Note for those changes, moved to hr_expense_sheet.py in 17.0 (don't hesitate when forward-ports arrive there to ask for help)

JulienAlardot commented on addons/hr_expense/models/hr_expense.py:
```diff
@@ -970,6 +996,12 @@ def _compute_attachment_number(self):
+    @api.depends('expense_line_ids.attachment_ids')
+    def _compute_main_attachment(self):
+        for sheet in self:
+            if not sheet.message_main_attachment_id or sheet.message_main_attachment_id not in sheet.attachment_ids:
+                sheet.message_main_attachment_id = sheet.attachment_ids[:1]
```
in 17.0 would change into something like
```suggestion
            if not sheet.message_main_attachment_id or sheet.message_main_attachment_id not in sheet.attachment_ids:
                sheet.message_main_attachment_id = sheet.expense_line_ids[:1].main_attachment_id
```

pivi-odoo replied:
Thanks for the heads up :)

JulienAlardot commented on addons/hr_expense/models/ir_attachment.py:
```diff
@@ -0,0 +1,39 @@
+            expenses_attachments = attachments.filtered(lambda att: att.res_model == 'hr.expense')
+            if expenses_attachments:
+                expenses = self.env['hr.expense'].browse(expenses_attachments.mapped('res_id'))
+                for expense in expenses.filtered(lambda exp: bool(exp.sheet_id)):
```
+1

pivi-odoo replied:
I always forget you can use filtered with a str of a field.
The bool was just add because of a linting warning, true it was not necessary.

JulienAlardot commented on addons/hr_expense/models/ir_attachment.py:
```diff
@@ -0,0 +1,39 @@
+            if expenses_attachments:
+                expenses = self.env['hr.expense'].browse(expenses_attachments.mapped('res_id'))
+                for expense in expenses.exists().filtered(lambda exp: bool(exp.sheet_id)):
+                    checksums = set(expense.attachment_ids.mapped('checksum'))
+                    attachments_to_unlink = expense.sheet_id.attachment_ids.filtered(lambda att: att.checksum in checksums)
```
+1

pivi-odoo replied:
It's true the pattern is recurrent, but it's never actually the same.
Sometimes checksums are based on some overlap or difference of the attachments that needs to be considered (to avoid touching attachments owned by the sheet, double copies, or deleting attachments on other expense lines).
Same argument for the attachments we filter on, sometimes it's on those of the expense, other time on those of the sheet.

IMHO If I were to refactor this pattern in a method, it would provide little value, since both the attachments we filter on, and the checksums we consider for the filtering would need to be injected as params, and the method itself wouldn't provide any value besides hiding a filtered, and maybe giving a "name" to the pattern.

Unless you had something specific else in mind that I am missing?

JulienAlardot commented on addons/hr_expense/tests/test_expenses.py:
```diff
@@ -1179,3 +1181,86 @@ def test_expense_from_attachments(self):
+            expense_with_attachment_1, expense_with_attachment_2, expense_with_attachment_3 = self.env['hr.expense'].create([{
+                'name': 'expense_1',
+                'employee_id': self.expense_employee.id,
+                'product_id': self.product_a.id,
+                'unit_amount': 1000.00,
+            }, {
+                'name': 'expense_2',
+                'employee_id': self.expense_employee.id,
+                'product_id': self.product_a.id,
+                'unit_amount': 999.00,
+            }, {
+                'name': 'expense_3',
+                'employee_id': self.expense_employee.id,
+                'product_id': self.product_a.id,
+                'unit_amount': 998.00,
+            }])
+            self.env['ir.attachment'].create([{
+                'name': "test_file_1.png",
+                'datas': base64.b64encode(b'content'),
+                'res_id': expense_with_attachment_1.id,
+                'res_model': 'hr.expense',
+            }, {
+                'name': "test_file_2.png",
+                'datas': base64.b64encode(b'other content'),
+                'res_id': expense_with_attachment_2.id,
+                'res_model': 'hr.expense',
+            }, {
+                'name': "test_file_3.png",
+                'datas': base64.b64encode(b'different content'),
+                'res_id': expense_with_attachment_3.id,
+                'res_model': 'hr.expense',
+            }])
+
+            expense_sheet = self.env['hr.expense.sheet'].create({
+                'company_id': self.env.company.id,
+                'employee_id': self.expense_employee.id,
+                'name': 'test sheet',
+                'expense_line_ids': [Command.set([expense_with_attachment_1.id, expense_with_attachment_2.id, expense_with_attachment_3.id])],
+            })
+
```
To avoid re-writing your tests at every forward-ports, in your expenses creation use 
```
'product_id': self.product_c.id,
'total_amount': X,
```
instead of using product_a and overriding the unit_amount (and using a default quantity of 1)

pivi-odoo replied:
Thanks for the info and avoiding me potentially useless conflicts :)

JulienAlardot reviewed (changes requested):
Sorry you had to touch pre-17.0 expense :p

Added a few comments, don't hesitate to ask for help in your forward ports
Thx for the changes

pivi-odoo commented:
I've addressed the comments above and pushed changes to implement the suggestions (namely the batching of unlink that I had forgotten to do).
Thanks for the heads up for 17.0 @JulienAlardot , I will probably pester you later on when the fw-port eventually reaches 17.0 :) to be inlined with the newer specs.

Note that I will be off until the end of the year, so I will be active on this PR once again next year.

(as of writing this the CI is still running, if it fails (shouldn't normally) I will take care of it later)

JulienVR commented on addons/hr_expense/tests/test_expenses.py:
```diff
@@ -1179,3 +1181,86 @@ def test_expense_from_attachments(self):
+        - When deleting an expense that is associated with an hr.expense.sheet
+        - When adding/removing an attachment of an expense that is associated with an hr.expense.sheet
+        """
+        def assert_attachments_are_synced(attachments_from_expenses, attachments_from_sheet, attachments_on_sheet, sheet_has_attachment):
```
this could simply be
```suggestion
        def assert_attachments_are_synced(sheet, attachments_on_sheet, sheet_has_attachment):
```

JulienVR commented on addons/hr_expense/tests/test_expenses.py:
```diff
@@ -1179,3 +1181,86 @@ def test_expense_from_attachments(self):
+        def assert_attachments_are_synced(attachments_from_expenses, attachments_from_sheet, attachments_on_sheet, sheet_has_attachment):
+            if sheet_has_attachment:
+                self.assertTrue(bool(attachments_on_sheet), "Attachment that belongs to the hr.expense.sheet only was removed unexpectedly")
+            self.assertSetEqual(set(attachments_from_expenses.mapped('checksum')), set((attachments_from_sheet - attachments_on_sheet).mapped('checksum')),
+                                "Attachments between expenses and their sheet is not in sync.")
```
And then, this becomes:
```suggestion
            self.assertSetEqual(
                set(sheet.expense_line_ids.attachment_ids.mapped('checksum')),
                set((sheet.attachment_ids - attachments_on_sheet).mapped('checksum')),
                "Attachments between expenses and their sheet is not in sync.",
            )
```

JulienVR commented on addons/hr_expense/tests/test_expenses.py:
```diff
@@ -1179,3 +1181,86 @@ def test_expense_from_attachments(self):
+                                "Attachments between expenses and their sheet is not in sync.")
+
+        for sheet_has_attachment in (False, True):
+            expense_with_attachment_1, expense_with_attachment_2, expense_with_attachment_3 = self.env['hr.expense'].create([{
```
it's easier to read if it's simply
```suggestion
            expense_1, expense_2, expense_3 = self.env['hr.expense'].create([{
```

JulienVR commented on addons/hr_expense/tests/test_expenses.py:
```diff
@@ -1179,3 +1181,86 @@ def test_expense_from_attachments(self):
+                'total_amount': 998,
+            }])
+            self.env['ir.attachment'].create([{
+                'name': "test_file_1.png",
```
let's avoid the annoying log `Post processing ignored : This file could not be decoded as an image file.`
by using a .txt file (or anything else actually, as long as the logs are not polluted :) )

pivi-odoo replied:
hmmm didn't see this line in my logs, odd, I've changed the extension to txt, should be ok?

JulienVR commented on addons/hr_expense/tests/test_expenses.py:
```diff
@@ -1179,3 +1181,86 @@ def test_expense_from_attachments(self):
+            assert_attachments_are_synced(expense_sheet.expense_line_ids.attachment_ids, expense_sheet.attachment_ids, sheet_attachment, sheet_has_attachment)
+            expense_with_attachment_2.sheet_id = expense_sheet
+            assert_attachments_are_synced(expense_sheet.expense_line_ids.attachment_ids, expense_sheet.attachment_ids, sheet_attachment, sheet_has_attachment)
+            expense_sheet.write({'expense_line_ids': [Command.set([expense_with_attachment_1.id, expense_with_attachment_3.id])]})
```
```suggestion
            sheet.expense_line_ids = [Command.set([expense_1.id, expense_3.id])]
```

JulienVR commented on addons/hr_expense/tests/test_expenses.py:
```diff
@@ -1179,3 +1181,86 @@ def test_expense_from_attachments(self):
+            assert_attachments_are_synced(expense_sheet.expense_line_ids.attachment_ids, expense_sheet.attachment_ids, sheet_attachment, sheet_has_attachment)
+            expense_with_attachment_3.unlink()
+            assert_attachments_are_synced(expense_sheet.expense_line_ids.attachment_ids, expense_sheet.attachment_ids, sheet_attachment, sheet_has_attachment)
+            expense_sheet.attachment_ids.filtered(lambda att: att.checksum in expense_sheet.expense_line_ids.attachment_ids.mapped('checksum')).unlink()
```
indent like this for instance:
```suggestion
            sheet.attachment_ids.filtered(
                lambda att: att.checksum in sheet.expense_line_ids.attachment_ids.mapped('checksum')
            ).unlink()
```

JulienVR commented on addons/hr_expense/tests/test_expenses.py:
```diff
@@ -1179,3 +1181,86 @@ def test_expense_from_attachments(self):
+                'res_model': 'hr.expense.sheet',
+            }) if sheet_has_attachment else self.env['ir.attachment']
+
+            assert_attachments_are_synced(expense_sheet.expense_line_ids.attachment_ids, expense_sheet.attachment_ids, sheet_attachment, sheet_has_attachment)
```
so this should be:
```suggestion
            assert_attachments_are_synced(sheet, sheet_attachment, sheet_has_attachment)
```

JulienVR commented on addons/hr_expense/models/hr_expense.py:
```diff
@@ -454,8 +458,35 @@ def write(self, vals):
+                previous_sheet = expense_to_previous_sheet[expense]
+                checksums = set((expense.attachment_ids - previous_sheet.expense_line_ids.attachment_ids).mapped('checksum'))
+                attachments_to_unlink += previous_sheet.attachment_ids.filtered(lambda att: att.checksum in checksums)
+                if vals['sheet_id']:
+                    if expense.sheet_id != previous_sheet:
```
```suggestion
                if vals['sheet_id'] and expense.sheet_id != previous_sheet:
```

JulienVR commented on addons/hr_expense/models/ir_attachment.py:
```diff
@@ -0,0 +1,38 @@
+from odoo import models, api
+
```
add a line here

JulienVR commented on addons/hr_expense/models/hr_expense.py:
```diff
@@ -454,8 +458,35 @@ def write(self, vals):
+            # The sheet_id has been modified, either by an explicit write on sheet_id of the expense,
+            # or by processing a command on the sheet's expense_line_ids.
+            # We need to delete the attachments on the previous sheet coming from the expenses that were modified,
+            # and copy the attachments of the expenses to the new sheet, if it's a no-op (writing same sheet_id as the current sheet_id of the expense),
```
indent please

JulienVR reviewed (commented):
My 2 cents :) 
Thanks for the change !
When you're done, you can ask again for @csnauwaert's review :)

pivi-odoo commented:
@JulienVR Thanks, I've pushed your suggestions.

PR is ready for review @csnauwaert :)

pivi-odoo commented:
Hello @csnauwaert, would you be kind enough to share your input on this PR when you have the time, thanks :)

csnauwaert commented:
Seems good to me. Thanks for the fix :)
@robodoo r+
